text/0000-io-net-2.1.md

```diff
@@ -0,0 +1,527 @@
+- Feature Name: `net2`
+- Start Date: 2015-05-07
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Expand the surface area of `std::net` to bind more low-level interfaces and
+provide more advanced customization and configuration of sockets.
+
+# Motivation
+
+Right now the API of `std::net` is fairly conservative in what it exposes, and
+there's quite a bit more functionality that can be exposed in a cross-platform
+manner. Some examples of tasks which cannot be accomplished through `std::net`
+APIs today are:
+
+* The high-level functions `TcpStream::connect`, `TcpListener::bind`, and
+  `UdpSocket::bind` actually encompass a few discrete steps as part of each
+  operation. Each step can possibly have configuration specified inbetween as
+  well as having extra parameters not specified to the top-level function. These
+  operations should be decomposable into independent operations.
+
+* There are a number of existing methods for setting the various options for a
+  socket, but they are far from comprehensive and haven't quite yet been
+  prepared for expansion. Many new socket options will likely be added to the
+  standard library over time so there needs to be room and design space for
+  these options.
+
+* A few remaining clarification issues remain around cross-platform
+  compatibility and behavior of APIs in various corner cases, and this RFC
+  attempts to clarify these areas.
+
+# Detailed design
+
+This RFC proposes APIs for each of the points listed in the motivation above,
+split into relevant sections.
+
+### std::net::TcpBuilder
+
+The following API is proposed to supplement the existing types in the `std::net`
+module:
+
+```rust
+pub struct TcpBuilder { ... }
+
+impl TcpBuilder {
+    /// Constructs a new TcpBuilder with either the `AF_INET` or `AF_INET6`
+    /// domain, the `SOCK_STREAM` type, and with a protocol argument of 0.
+    ///
+    /// Note that passing other kinds of flags or arguments can be done through
+    /// the `FromRaw{Fd,Socket}` implementation.
+    pub fn new_v4() -> io::Result<TcpBuilder>;
+    pub fn new_v6() -> io::Result<TcpBuilder>;
+
+    /// Binds this socket to the specified address.
+    ///
+    /// This function directly corresponds to the bind(2) function on Windows
+    /// and Unix.
+    pub fn bind<T: ToSocketAddrs>(&self, addr: T) -> io::Result<&TcpBuilder>;
```

Comment:
It should probably be noted in the docs that using this in a builder-like fashion isn't possible in a non-painful fashion now (due to the `io::Result` wrapper), but may be in the future with something like `?` sugar.